File: rust/566.reshape-the-matrix.rs
/*
 * @lc app=leetcode id=566 lang=rust
 *
 * [566] Reshape the Matrix
 */

// @lc code=start
impl Solution {
    pub fn matrix_reshape(mat: Vec<Vec<i32>>, r: i32, c: i32) -> Vec<Vec<i32>> {
        let m = mat.len();
        let n = mat[0].len();
        let (r, c) = (r as usize, c as usize);
        if m * n != r * c || (m == r && n == c) {
            return mat;
        }
        mat.into_iter()
            .flatten()
            .collect::<Vec<i32>>()
            .chunks(c)
            .map(|row| row.to_vec())
            .collect::<Vec<_>>()
    }
}
// @lc code=end

File: javascript/566.reshape-the-matrix.js
/*
 * @lc app=leetcode id=566 lang=javascript
 *
 * [566] Reshape the Matrix
 */

// @lc code=start
/**
 * @param {number[][]} mat
 * @param {number} r
 * @param {number} c
 * @return {number[][]}
 */
var matrixReshape = function (mat, r, c) {
  const m = mat.length,
    n = mat[0].length;
  if (m * n !== r * c || (m === r && n === c)) return mat;
  return mat.flat().reduce((acc, num, i) => {
    if (i % c === 0) {
      acc.push([]);
    }
    acc[acc.length - 1].push(num);
    return acc;
  }, []);
};
// @lc code=end
